opensearch-operator/controllers/clusterController.go

```diff
@@ -4,122 +4,78 @@ import (
-//+kubebuilder:rbac:groups=opensearch.opster.io,resources=opensearchcluster/finalizers,verbs=update
+func (r *ClusterReconciler) Reconcile(controllerContext *ControllerContext) (*opsterv1.ComponentStatus, error) {
 
-func (r *ClusterReconciler) Reconcile(context.Context, ctrl.Request) (ctrl.Result, error) {
-
-	cm := v1.ConfigMap{}
 	namespace := r.Instance.Spec.General.ClusterName
 
-	cmName := "opensearch-yml"
-
-	if err := r.Get(context.TODO(), client.ObjectKey{Name: cmName, Namespace: namespace}, &cm); err != nil {
-		clusterCm := builders.NewCmForCR(r.Instance)
-		err = r.Create(context.TODO(), clusterCm)
-		if err != nil {
-			if !errors.IsAlreadyExists(err) {
-				fmt.Println(err, "Cannot create Configmap "+clusterCm.Name)
-				r.Recorder.Event(r.Instance, "Warning", "Cannot create Configmap ", "Requeue - Fix the problem you have on main Opensearch ConfigMap")
-				return ctrl.Result{}, err
-			}
-		}
-		fmt.Println("Cm Created successfully", "name", clusterCm.Name)
-	}
-
-	headlessService := v1.Service{}
-	serviceName := r.Instance.Spec.General.ServiceName + "-headless-service"
-	if err := r.Get(context.TODO(), client.ObjectKey{Name: serviceName, Namespace: namespace}, &headlessService); err != nil {
-		/// ------ Create Headless Service -------
-		headless_service := builders.NewHeadlessServiceForCR(r.Instance)
-		err = r.Create(context.TODO(), headless_service)
-		if err != nil {
-			if !errors.IsAlreadyExists(err) {
-				fmt.Println(err, "Cannot create Headless Service")
-				r.Recorder.Event(r.Instance, "Warning", "Cannot create Headless Service ", "Requeue - Fix the problem you have on main Opensearch Headless Service ")
-
-				return ctrl.Result{}, err
-			}
-		}
-		fmt.Println("service Created successfully", "name", headless_service.Name)
-	}
-
 	service := v1.Service{}
-	serviceName = r.Instance.Spec.General.ServiceName + "-svc"
+	serviceName := r.Instance.Spec.General.ServiceName
 	if err := r.Get(context.TODO(), client.ObjectKey{Name: serviceName, Namespace: namespace}, &service); err != nil {
-
-		/// ------ Create External Service -------
+		// Create External Service
 		clusterService := builders.NewServiceForCR(r.Instance)
 
 		err = r.Create(context.TODO(), clusterService)
 		if err != nil {
 			if !errors.IsAlreadyExists(err) {
-				fmt.Println(err, "Cannot create service")
+				r.Logger.Error(err, "Cannot create service")
 				r.Recorder.Event(r.Instance, "Warning", "Cannot create opensearch Service ", "Requeue - Fix the problem you have on main Opensearc Service ")
-				return ctrl.Result{}, err
+				return nil, err
 			}
 
 		}
-		fmt.Println("service Created successfully", "name", service.Name)
-
+		r.Logger.Info(fmt.Sprintf("service %s created successfully", clusterService.Name))
 	}
 
-	///// ------ Create Es Nodes StatefulSet -------
-	NodesCount := len(r.Instance.Spec.NodePools)
-	sts := sts.StatefulSet{}
+	// Create StatefulSets for NodePools
```

Comment:
Why headless service is created for every nodePool?

Reply:
As stated in the [kubernetes docs](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) every StatefulSet needs a headless service (`StatefulSets currently require a Headless service to be responsible for the network identity of the Pods. You are responsible for creating this Service.`. As we will need this network identity for per-node certificates I already added it here.

Reply:
I familiar with only one headless service for the masters discovery. can you explain that again `As we will need this network identity for per-node certificates ?`

Reply:
With a headless service defined for a STS each pod gets a DNS-resolvable name (e.g. master-0.my-cluster.my-cluster.svc.cluster.local) and can be directly reached. The opensearch nodes can then advertise this hostname instead of their IP to the other nodes for node-2-node communication. This helps with per-pod certificates because we can then create certificates for the pod hostname (e.g. master-0) which stays the same instead of for the pod IP which will change when the pod is recreated (e.g. to change cpu/ram resources). And the certificate needs the hostname or pod IP so the other nodes can do correct TLS connection certificate validation.

Reply:
Understood, thanks!